### PR TITLE
extend syntax on fir.do_loop to pass back the final value.

### PR DESCRIPTION
**Pull request by schweitzpgi**
extend syntax for fir.iterate_while as well.

### CODE REVIEW

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 1]
```diff
@@ -608,32 +608,59 @@ static mlir::ParseResult parseIterWhileOp(mlir::OpAsmParser &parser,
...
       if (parser.resolveOperand(std::get<0>(operand_type),
                                 std::get<1>(operand_type), result.operands))
-        return mlir::failure();
-    result.addTypes(regionTypes);
+        return failure();
+    if (hasResult) {
+      result.addTypes(regionTypes);
+    } else {
+      result.addTypes(i1Type);
+      result.addTypes(resTypes);
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:637]: There are actually two distinct types. There is the result types of the Op itself, and there are the internal region type of the code embedded in the Op. These are similar but not the same.

File: flang/test/Fir/loop01.fir [hunk 2]
```diff
@@ -20,3 +20,162 @@ func @x(%lb : index, %ub : index, %step : index, %b : i1, %addr : !fir.ref<index
...
+  %c1 = constant 1 : index
+  %ok1 = constant true
+  // CHECK: %[[ok4:.*]] = call i1 @f2()
+  %ok4 = fir.call @f2() : () -> i1
+  // CHECK-DAG: %[[count:.*]] = phi i64
+  // CHECK-DAG: %[[ok3:.*]] = phi i1 {{.*}}[ true
+  // CHECK-DAG: %[[j:.*]] = phi i1 {{.*}}[ %[[ok4]]
+  // CHECK: %[[prev:.*]] = icmp slt i64 %[[count]],
+  // CHECK: = and i1 %[[prev]], %[[ok3]]
+  %ok2:2 = fir.iterate_while (%i = %lo to %up step %c1) and (%ok3 = %ok1) iter_args(%j = %ok4) -> i1 {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/test/Fir/loop01.fir:73]: This maintains conformance to the pre-existing syntax. The early exit boolean was, is, and will be returned always. The return type here is for the iter_args.

The new extended syntax let's one explicitly add the counter and the early exit condition both.

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 3]
```diff
@@ -608,32 +608,59 @@ static mlir::ParseResult parseIterWhileOp(mlir::OpAsmParser &parser,
...
     llvm::SmallVector<mlir::Type, 4> regionTypes;
     // Parse assignment list and results type list.
     if (parser.parseAssignmentList(regionArgs, operands) ||
         parser.parseArrowTypeList(regionTypes))
-      return mlir::failure();
+      return failure();
+    if (regionTypes.size() == operands.size() + 2)
+      hasResult = true;
+    llvm::ArrayRef<mlir::Type> resTypes = regionTypes;
+    resTypes = hasResult ? resTypes.drop_front(2) : resTypes;
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:627]: Are you asking if the test in function `@x2` compiles?

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Optimizer/Dialect/FIROps.cpp:627]: The verifier already checks the types.

File: flang/test/Fir/loop01.fir (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/test/Fir/loop01.fir:73]: I did not pay attention the pre-existing syntax was to hide the boolean result. What is the point of hiding it though ?
I feel that for an mlir code reader not 100% familiar with fir its weird to have `%ok:2 = op -> i1`. It's then hard to guess what are `ok#1` and `ok#2` types from code reading.

File: flang/lib/Optimizer/Dialect/FIROps.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Optimizer/Dialect/FIROps.cpp]: OK, thanks for clarifying.

**Review by jeanPerier — APPROVED** [review]

File: flang/lib/Optimizer/Dialect/FIROps.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Optimizer/Dialect/FIROps.cpp]: No, I thought that the `-> i1` in `x3` referred to both the boolean result and the returned arguments (that happened to also be a boolean). So I was wondering what happened if iter_args was not an i1 in this case. But you explained below that the boolean type is not printed in the result type list, so my thinking was wrong.

File: flang/test/Fir/loop01.fir (hunk 2 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/test/Fir/loop01.fir:73]: I might change this, but for now I wanted to minimize the churn for churn's sake.

**Pull request merged**